resteasy-client-reactor-netty/src/main/java/org/jboss/resteasy/client/jaxrs/engines/ReactorNettyClientHttpEngine.java

```diff
@@ -196,6 +169,81 @@ public ReactorNettyClientHttpEngine(final HttpClient httpClient,
                 });
     }
 
+    /**
+     * The main business logic mapping RestEasy's {@link ClientInvocation request} to Reactor Netty's concept
+     * of it.
+     */
+    private HttpClient.ResponseReceiver<?> send(final ClientInvocation resteasyReq) {
+        final Optional<Object> reqPayload = Optional.ofNullable(resteasyReq.getEntity());
+
+        final HttpClient.RequestSender requestSender = httpClient.headers(headers -> addHeaders(resteasyReq, headers))
+                    .request(HttpMethod.valueOf(resteasyReq.getMethod()))
+                    .uri(resteasyReq.getUri().toString());
+
+        return reqPayload.<HttpClient.ResponseReceiver<?>>map(ignore -> {
+            return requestSender.send((reactorReq, outbound) -> {
+                final ByteBufOutputStream byteBufOutputStream =
+                    new ByteBufOutputStream(outbound.alloc().buffer());
+
+                /* Replacing the ClientRequestHeaders with TrackingClientRequestHeaders
+                to track the changes to headers by the sendRequestBody method. */
+                resteasyReq.setHeaders(
+                    new TrackingClientRequestHeaders(
+                        resteasyReq.getClientConfiguration(),
+                        resteasyReq.getHeaders().getHeaders()
+                    )
+                );
+
+                try {
+                    sendRequestBody(resteasyReq, byteBufOutputStream);
+                } catch (final IOException ioe) {
+                    return Mono.error(ioe);
+                }
+
+                // Updating the HttpClientRequest with the headers modified by the sendRequestBody method.
+                final TrackingMap<?> trackingMap = (TrackingMap<?>) resteasyReq.getHeaders().getHeaders();
+                trackingMap.getAddedOrUpdatedKeys()
+                    .forEach(key -> updateHeader(
+                            key,
+                            resteasyReq.getHeaders().getHeaders(),
+                            reactorReq.requestHeaders()
+                        )
+                    );
+                trackingMap.getRemovedKeys()
+                    .forEach(reactorReq.requestHeaders()::remove);
```

Comment:
I'm not really user I follow this. Could you explain the reasoning a bit?

Reply:
Ultimately, we need to get all the headers established in the RestEasy client request into ReactorNetty's client request.  One of our first steps is to do this copy.  However, as we (painfully) found out, the `MessageBodyWriter` can add/remove/update headers.  So we have this 'tracking structure' that  @dnagarajan89 wrote to make adjustments to headers as a final step.

Reply:
Okay got it. I think I'd rather see `TrackingClientRequestHeaders` and `TrackingMap` defined in this module rather than the `resteasy-client` and `resteasy-core` modules. Just keeps it more contained and doesn't make the public API for now :)

Reply:
@jamezp we addressed your suggestion above on moving the `TrackingClientRequestHeaders` and `TrackingMap` to reactor-netty-client-integration module.